src/v/archival/archiver_data_upload_workflow.cc

```diff
@@ -0,0 +1,770 @@
+                    .ntp = fsm._ntp,
+                  });
+                  break;
+              };
+          });
+    }
+
+    ss::future<> shutdown(state_machine_t& fsm) {
+        vassert(!_as.abort_requested(), "shutdown is called twice");
+        _as.request_abort();
+        auto& st = fsm.get_state<st_terminal&>();
+        vlog(_log.info, "Waiting for the workflow shutdown");
+        return st.completed_shutdown.get_future();
+    }
+
+    explicit data_upload_workflow_fsm(
+      ss::shared_ptr<archiver_operations_api> api,
+      ss::shared_ptr<archiver_scheduler_api> quota_api,
+      model::ktp ntp,
+      model::term_id term)
+      : _ops_api(std::move(api))
+      , _quota_api(std::move(quota_api))
+      , _ntp(std::move(ntp))
+      , _term(term)
+      , _rtc(_as)
+      , _log(
+          archival_log, _rtc, ssx::sformat("{}-{}", ntp.to_ntp().path(), term))
+      , _local_segment_size(config::shard_local_cfg().log_segment_size.bind())
+      , _target_segment_size(
+          config::shard_local_cfg().cloud_storage_segment_size_target.bind())
+      , _min_segment_size(
+          config::shard_local_cfg().cloud_storage_segment_size_min.bind())
+      , _segment_upload_timeout(
+          config::shard_local_cfg()
+            .cloud_storage_segment_upload_timeout_ms.bind())
+      , _manifest_upload_timeout(
+          config::shard_local_cfg()
+            .cloud_storage_manifest_upload_timeout_ms.bind())
+      , _metadata_sync_timeout(config::shard_local_cfg()
+                                 .cloud_storage_metadata_sync_timeout_ms.bind())
+      , _upload_backoff(
+          config::shard_local_cfg().cloud_storage_initial_backoff_ms.bind())
+      , _upload_interval(
+          config::shard_local_cfg()
+            .cloud_storage_segment_max_upload_interval_sec.bind())
+      , _compacted_reuploads_enabled(
+          config::shard_local_cfg()
+            .cloud_storage_enable_compacted_topic_reupload.bind()) {}
+
+    // Default no-transition handler.
+    // template<class FSM, class Event>
+    // void no_transition(const Event&, FSM&, int) {}
+
+    // Default exception handler.
+    template<class FSM, class Event>
+    void exception_caught(const Event&, FSM&, std::exception& err) {
+        vassert(false, "Unexpected exception {}", err);
+    }
+
+private:
+    std::pair<size_t, size_t> get_segment_size_bounds() {
+        auto high = _target_segment_size().value_or(_local_segment_size());
+        auto low = _min_segment_size().value_or(high / 2);
+        if (low > high) {
+            low = high * 8 / 10;
+        }
+        return std::make_pair(low, high);
+    }
+
+    ss::shared_ptr<archiver_operations_api> _ops_api;
+    ss::shared_ptr<archiver_scheduler_api> _quota_api;
+    model::ktp _ntp;
+    model::term_id _term;
+    ss::abort_source _as;
+    retry_chain_node _rtc;
+    retry_chain_logger _log;
+    config::binding<size_t> _local_segment_size;
```

Comment:
nit: encapsulate all `config::binding` members in a new `archiver_configuration` object?